pmlc/dialect/pxa/transforms/tileaccumlate.cc

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Intel Corporation
+
+#include <limits>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "mlir/Analysis/AffineStructures.h"
+#include "mlir/Pass/Pass.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "pmlc/dialect/pxa/transforms/tile.h"
+
+#include "mlir/Support/DebugStringHelper.h"
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/analysis/uses.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+using mlir::AffineParallelOp;
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+using llvm::DenseMap;
+using llvm::DenseSet;
+using llvm::SmallVector;
+using mlir::BlockArgument;
+
+Operation *GetOriginalDef(Value val) {
+  auto opRes = val.cast<mlir::OpResult>();
+  while (true) {
+    auto ap = mlir::dyn_cast<AffineParallelOp>(opRes.getOwner());
+    if (!ap)
+      break;
+    auto ret = mlir::cast<AffineYieldOp>(ap.getBody()->getTerminator());
+    auto src = ret.getOperand(opRes.getResultNumber());
+    auto defop = src.getDefiningOp();
+    if (dyn_cast<AffineReduceOp>(defop)) {
+      opRes = src.cast<mlir::OpResult>();
+    } else if (dyn_cast<AffineIfOp>(defop)) {
+      defop->walk([&](AffineReduceOp op) {
+        opRes = op.getResult().cast<mlir::OpResult>();
+      });
+    }
+  }
+  return opRes.getOwner();
+}
+
+bool isAggregation(AffineParallelOp op) {
```

Comment:
Currently a few test examples work as expected. Testing and checking with more complex tests.